File: cpp_examples/customized_http/src/customized_converter.cc
/*
 * Copyright 2022 netease. All rights reserved.
 * Author zhaochaochao@corp.netease.com
 * Date   2023/02/08
 * Brief  Customized converter of model, including pre-process and post-process.
 */

#include "customized_converter.h"

#include <rapidjson/document.h>
#include <rapidjson/stringbuffer.h>
#include <rapidjson/writer.h>

#include "logger/logger.h"

namespace netease::grps {

YourConverter::YourConverter() = default;
YourConverter::~YourConverter() = default;

void YourConverter::Init(const std::string& path, const YAML::Node& args) {
  Converter::Init(path, args);
  // Add your codes here.
  CLOG4(INFO, "your converter init, path: " << path << ", args: " << args);
}

void YourConverter::PreProcess(const ::grps::protos::v1::GrpsMessage& input,
                               std::vector<std::pair<std::string, TensorWrapper>>& output,
                               GrpsContext& context) {
  auto* cntl = context.http_controller();

  // Check content type.
  const auto& content_type = cntl->http_request().content_type();
  if (content_type != "application/json") {
    // context.set_has_err(true);
    context.set_err_msg("Unsupported media type: " + content_type);
    cntl->http_response().set_status_code(brpc::HTTP_STATUS_BAD_REQUEST);
    cntl->http_response().set_content_type("text/plain");
    cntl->response_attachment().append(context.err_msg());
    CLOG4(ERROR, "YourConverter::PreProcess failed, err msg: " << context.err_msg());
    return;
  }

  // Parse "a" and "b" from json.
  rapidjson::Document doc;
  doc.Parse(cntl->request_attachment().to_string().c_str());
  if (doc.HasParseError()) {
    // context.set_has_err(true);
    context.set_err_msg("Parse json failed.");
    cntl->http_response().set_status_code(brpc::HTTP_STATUS_BAD_REQUEST);
    cntl->http_response().set_content_type("text/plain");
    cntl->response_attachment().append(context.err_msg());
    CLOG4(ERROR, "YourConverter::PreProcess failed, err msg: " << context.err_msg());
    return;
  }

  if (!doc.HasMember("a") || !doc.HasMember("b")) {
    // context.set_has_err(true);
    context.set_err_msg("Json format error.");
    cntl->http_response().set_status_code(brpc::HTTP_STATUS_BAD_REQUEST);
    cntl->http_response().set_content_type("text/plain");
    cntl->response_attachment().append(context.err_msg());
    CLOG4(ERROR, "YourConverter::PreProcess failed, err msg: " << context.err_msg());
    return;
  }
  auto a = doc["a"].GetFloat();
  auto b = doc["b"].GetFloat();

  TensorWrapper a_tensor(a);
  TensorWrapper b_tensor(b);
  output.emplace_back("a", std::move(a_tensor));
  output.emplace_back("b", std::move(b_tensor));
}

void YourConverter::PostProcess(const std::vector<std::pair<std::string, TensorWrapper>>& input,
                                ::grps::protos::v1::GrpsMessage& output,
                                GrpsContext& context) {
  auto* cntl = context.http_controller();

  if (input.empty()) {
    // context.set_has_err(true);
    context.set_err_msg("Infer out is empty.");
    cntl->http_response().set_status_code(brpc::HTTP_STATUS_INTERNAL_SERVER_ERROR);
    cntl->http_response().set_content_type("text/plain");
    cntl->response_attachment().append(context.err_msg());
    CLOG4(ERROR, "YourConverter::PostProcess failed, err msg: " << context.err_msg());
    return;
  }

  // Build response json body.
  auto c = input[0].second.float_tensor;
  rapidjson::Document doc;
  doc.SetObject();
  doc.AddMember("c", c, doc.GetAllocator());
  rapidjson::StringBuffer buffer;
  rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
  doc.Accept(writer);

  // Set response.
  cntl->http_response().set_status_code(brpc::HTTP_STATUS_OK);
  cntl->http_response().set_content_type("application/json");
  cntl->response_attachment().append(buffer.GetString());
}

// Implement and register other converter class after here.

} // namespace netease::grps

File: py_examples/customized_http/src/customized_converter.py
# Copyright 2022 netease. All rights reserved.
# Author zhaochaochao@corp.netease.com
# Date   2023/9/5
# Brief  Customized converter of model, including pre-process and post-process.
import http
import json
import traceback

from flask import Response
from grps_framework.apis.grps_pb2 import GrpsMessage
from grps_framework.context.context import GrpsContext
from grps_framework.converter.converter import Converter, converter_register
from grps_framework.logger.logger import clogger


class YourConverter(Converter):
    """Your converter."""

    def init(self, path=None, args=None):
        """
        Init converter.

        Args:
            path: Path.
            args: More args.

        Raises:
            Exception: If init failed, can raise exception and exception will be caught by server and show error message
            to user when start service.
        """
        super().init(path, args)
        clogger.info('your converter init, path: {}, args: {}'.format(path, args))

    def preprocess(self, inp: GrpsMessage, context: GrpsContext):
        """
        Preprocess.

        Args:
            inp: Input message from client or previous model(multi model sequential mode).
            context: Grps context of current request.

        Returns:
            Pre-processed data which is input of model inferer.

        Raises:
            Exception: If preprocess failed, can raise exception and exception will be caught by server and return error
            message to client.
        """
        try:
            clogger.info('your converter preprocess.')
            request = context.get_http_request()
            if request.content_type != 'application/json':
                raise Exception('content_type error, unsupported type: {}'.format(request.content_type))

            request_json = request.get_json()
            if 'a' not in request_json or 'b' not in request_json:
                raise Exception('request json error, need key a and b, request_json: {}'.format(request_json))

            a = float(request_json['a'])
            b = float(request_json['b'])
            return {'a': a, 'b': b}
        except Exception as e:
            error_message = traceback.format_exc()
            # context.set_has_err(True)
            context.set_err_msg(error_message)
            context.set_http_response(Response(response=error_message, status=http.HTTPStatus.INTERNAL_SERVER_ERROR,
                                               content_type="text/plain"))
            clogger.error('your inferer infer error: {}'.format(error_message))

        return {}

    def postprocess(self, inp, context: GrpsContext) -> GrpsMessage:
        """
        Postprocess.

        Args:
            inp: Input to be post-processed, which is output of model inferer.
            context: Grps context of current request.

        Returns:
            Post-processed data with GrpsMessage format to client or next model(multi model sequential mode).

        Raises:
            Exception: If postprocess failed, can raise exception and exception will be caught by server and return error
            message to client.
        """
        try:
            clogger.info('your converter postprocess.')
            if len(inp) == 0 or 'c' not in inp:
                raise Exception('infer output error, need key c, inp: {}'.format(inp))

            context.set_http_response((json.dumps(inp), http.HTTPStatus.OK, {"Content-Type": "application/json"}))
        except Exception as e:
            error_message = traceback.format_exc()
            # context.set_has_err(True)
            context.set_err_msg(error_message)
            context.set_http_response(
                (error_message, http.HTTPStatus.INTERNAL_SERVER_ERROR, {"Content-Type": "text/plain"}))
            clogger.error('your inferer infer error: {}'.format(error_message))
        return GrpsMessage()


converter_register.register('your_converter', YourConverter())
